### PR TITLE
[CI] Add a clear deadline

### PR DESCRIPTION
**Pull request by proux01**
It seems there was some agreement during last Coq call about that point so let's do it.
I'm personally a bit wary that 7 days might be a bit too demanding on CI entries developers but why not.

### CODE REVIEW

File: dev/ci/README-developers.md
Comment (SkySkimmer) [inline dev/ci/README-developers.md:60]: It is still ok to ask the project maintainer for help for plugins.

**Comment by SkySkimmer** [comment]
This PR seems to change more than just adding a deadline.

**Comment by proux01** [comment]
> This PR seems to change more than just adding a deadline.

The two other commits are minor doc updates, not sure they are worth their own PR?

File: dev/ci/README-developers.md
Comment (proux01) [reply dev/ci/README-developers.md:54]: Indeed, added a comment

**Comment by proux01** [comment]
I did put the unrelated minor changes in their own PR #19563

**Comment by andres-erbsen** [comment]
I'm generally in agreement with the proposed change. A slight variant that I have been advocating would be as follows:

1. The PR author writes a changelog entry that includes porting guidance.
2. The PR author validates that porting guidance on some representative relevant developments.
3. The assignee asks downstreams to port following the guidance. If it is non-obvious how to port with the guidance given., that is to be discussed on the PR.

What I am trying to emphasize here is that a relatively short timeline can only be expected from a relatively straightforward task, e.g. "follow these steps" as opposed to "figure out how to make it work".

**Comment by proux01** [comment]
Thanks, the clarification makes a lot of sense. I tried to incorporate it in the PR but feel free to further edit.

File: dev/ci/README-users.md
Comment (Zimmi48) [inline dev/ci/README-users.md:9]: It can happen that patches are sent for libraries as well, so maybe change to:

```suggestion
  developers and contributors will send you patches to adapt it (systematically
  only for plugins) or
```

File: dev/ci/README-developers.md
Comment (Zimmi48) [inline dev/ci/README-developers.md:60]: This looks good, but since this is long and it describes a well-defined process, I would encourage the use of bullet-points here.

Furthermore, this doesn't describe what "legitimate breakage" means, besides giving an example of the most obvious case. But many things beyond removing an already deprecated feature can be legitimate (and OTOH removing a recently deprecated but highly used feature might still be debatable).

So I would suggest clarifying that what counts as a legitimate breakage is something to be decided collectively (author, reviewers, assignee) based on the preliminary porting assessment made by the author.

**Review by Zimmi48 — APPROVED** [review]
This is a much welcome improvement, especially after incorporating Andres' feedback.

File: dev/ci/README-developers.md
Comment (proux01) [reply dev/ci/README-developers.md]: I agree, I gave it an attempt following your directions.

File: dev/ci/README-users.md
Comment (andres-erbsen) [inline dev/ci/README-users.md:22]:
```suggestion
  required by Coq developers or to integrate their patches (in 7 days for simple requests).
```

**Review by andres-erbsen — COMMENTED** [review]
I think we've basically got it down. I left some copyediting suggestions trying to spin the text in a way that reads better to me, but at this point it's just picking nits.

**Comment by JasonGross** [comment]
Can we also encourage PR writers to include automation instructions when possible (e.g., "use this sed script" or "use this python script" or "set this option, run coqc on your development and capture the logs, then feed them to this python script to update your files"), especially for PRs that induce many lines of changed code?

File: dev/ci/README-developers.md
Comment (Zimmi48) [inline dev/ci/README-developers.md:61]:
```suggestion
   determined with the PR assignee and reviewers, or discussed
```

File: dev/ci/README-developers.md
Comment (Zimmi48) [inline dev/ci/README-developers.md:72]: This is where we could mention the possibility of providing a porting scripts when many lines need to be updated (as per Jason's suggestion).

File: dev/ci/README-developers.md
Comment (Zimmi48) [inline dev/ci/README-developers.md:79]: The idea of providing a script is already there, but maybe it's a bit buried, since it's in a middle of a message template.

File: dev/ci/README-users.md
Comment (Zimmi48) [reply dev/ci/README-users.md:22]: This is changing the meaning of the policy, and we don't know who is to judge what is a simple request, so why not rather say:
```suggestion
  required by Coq developers or to integrate their patches (in a 7 days timeframe,
  unless you request more time or Coq developers give you longer, e.g., in case
  of more complex changes).
```

Comment (proux01) [reply dev/ci/README-users.md]: I agree, I tried something.

File: dev/ci/README-developers.md
Comment (proux01) [reply dev/ci/README-developers.md:75]: I rather put it into the previous item (mentioning porting instructions).

File: dev/ci/README-developers.md
Comment (proux01) [reply dev/ci/README-developers.md]: Yes, these were indeed somewhat misplaced. I put them earlier in the text body. That should answer @JasonGross 's comment.

File: dev/ci/README-developers.md
Comment (Zimmi48) [inline dev/ci/README-developers.md:68]:
```suggestion
   script to help porting if you wish.
```

**Review by Zimmi48 — APPROVED** [review]
Only one typo left. Let's not keep this waiting longer. I'll merge as soon as this is fixed.

**Comment by ejgallego** [comment]
> Only one typo left. Let's not keep this waiting longer. I'll merge as soon as this is fixed.

I am not sure this PR is so urgent, but IIUC this is a significant change for the CI policy and a move apart towards the original design goals of the CI.

With my @coq/ci-maintainers hat on, I think the state of this PR is premature and many essential questions have not been addressed, like what is an essential change. I'd recommend further discussion on CI policy and scenarios.

In particular, the change of policy poses a very serious risk of going towards a Coq 8.5 situation, where breakage crept in the 8.4 to 8.5 release, and only huge efforts by students and post-docs managed to fix all the broken developments.

The cost of updating a proof development should not be under estimated.

With the current policy and deadline, it seems to me that many projects we have in the CI will just fail to adapt, so this will lead to considerable divergences.

File: dev/ci/README-developers.md
Comment (proux01) [reply dev/ci/README-developers.md]: Fixed

**Comment by ejgallego** [comment]
For example, if I am not mistaken, @proux01 mentioned recently that CompCert did not meet the requirements for Coq CI.

What does this new policy mean for CompCert?

**Comment by andres-erbsen** [comment]
> Can we also encourage PR writers to include automation instructions when possible (e.g., "use this sed script" or "use this python script" or "set this option, run coqc on your development and capture the logs, then feed them to this python script to update your files"), especially for PRs that induce many lines of changed code?

I am very much in favor of doing this when possible. IMO we should be asking downstreams to be responsive and to understand their own code and to apply that understanding, but we should not be asking them to understand each change we want to make on a short notice. I've been trying to gesture at that with the language I've suggested for the document, and now I think there's indeed a policy question there. So, for example, I'd absolutely call time on any development (even CompCert) for merging a change we prepare, or say porting over a stdlib-definition renaming that breaks some Ltac2 code that dynamically constructs and dereferences definition names. I think it would loos us a lot of goodwill to tell any development to figure out how to get their code working again after we remove some unique functionality from Coq or stdlib. Somewhere between there should be a line, and I don't know how to put a finger on it.

> In particular, the change of policy poses a very serious risk of going towards a Coq 8.5 situation, where breakage crept in the 8.4 to 8.5 release, and only huge efforts by students and post-docs managed to fix all the broken developments.

When you imagine the increased risk, do you think of it as happening without us knowing (a seemingly simple change is actually broadly breaking, and we miss it because the CI is smaller than it would have been without this policy), or because a change is considered so valuable that it is worth pushing it through even though multiple CI users are not ported in time? IMO the policy being codified here should absolutely not be used for the latter: if it's not clear CI developments can be ported over a change, we should not make that change. And if the difficulties are missed during review and the change is erroneously merged, we should revert. Maybe I'm naive, but my sense of 8.5 is that similar attitude could have changed course there.

> With the current policy and deadline, it seems to me that many projects we have in the CI will just fail to adapt, so this will lead to considerable divergences.

I do not want to lose the majority of the CI. I am hoping it wouldn't happen, but I am very willing to thread carefully here. One option would be to start monitoring this policy without enforcing it: go through the motions, but instead of removing a development from CI, fix it ourselves and delay the PR until they merge as we have so far. Then in some months time, tally up the work, delays, and retained developments, and make a more informed decision. But we should only take this detour if we can't find a policy now for which we all have similar and acceptable predictions.

**Comment by ejgallego** [comment]
> When you imagine the increased risk, do you think of it as happening without us knowing (a seemingly simple change is actually broadly breaking, and we miss it because the CI is smaller than it would have been without this policy), or because a change is considered so valuable that it is worth pushing it through even though multiple CI users are not ported in time? IMO the policy being codified here should absolutely not be used for the latter: if it's not clear CI developments can be ported over a change, we should not make that change. And if the difficulties are missed during review and the change is erroneously merged, we should revert. Maybe I'm naive, but my sense of 8.5 is that similar attitude could have changed course there.

Thanks for your insights @andres-erbsen , the main risk for me is that the cost of reconciliation tends to be exponential on the divergence time (measured experimentally).

This metric means that in practice, most industry uses a mono-repos, where Delta = 0, that is to say, the repos is always coherent.

You make a very good point about the two possibilties, I am getting a feeling that folks want to change the policy to indeed make the development of Coq more agile. On the other hand, we risk a Compcert 8.5 or mathlib3 vs mathlib4 situation.

IMHO this is a discussion we should have with downstream developers.

Also, I feel like the importance of tooling in all these recent proposals has been minimal, however we have evidence in the larger industrial context that tooling is key: proper tooling can support way more agile dev setups.

In this sense, we have pretty advanced tooling ideas that will allow for example to do something like Lean 5 without breaking mathlib.

**Comment by proux01** [comment]
This PR is based on the assumption that we trust PR assignees / the Coq dev team to not have excessive demands toward CI project maintainers. So either:
1. We trust them and the PR should be good.
2. We don't and then we need a clear definition of what are acceptable demands to CI projects. This obviously makes the PR much more complicated.

My personal opinion is that none of use wants to kill the CI and it should be fine trusting assignees. In worst case, the 7 days delay will hopefully be enough to react and rollback before merging a problematic PR. At least, It's worth trying.

However I start realizing I may have been very optimistic by allowing the subject only 15 minutes in next Coq call: https://github.com/coq/coq/wiki/Coq-Call-2024-09-24

**Comment by ejgallego** [comment]
@proux01 , indeed, I think this topic is complex, and will require some time to get right. I (very well) understand how frustrating can be when things move slow, but in this case it cannot hurt to us to discuss a bit more.

In particular, it seems to me, if I understand things correctly, that we should involve some key Coq users in this discussion. In the end, the Coq team and its users have to work together with regards to compatibility.

**Comment by ppedrot** [comment]
@proux01 about the next call, note that this week is the Gallinette yearly retreat so most likely none of the people from there will be attending the call. If possible I'd recommend postponing the topic...

**Comment by proux01** [comment]
Thanks for pointing this out, I moved it to the next available slot (October 8th).

**Comment by Zimmi48** [comment]
Let's indeed discuss this again in a Coq Call. There was indeed no urgency to merge this PR. I thought it was the simple application of something that had already been decided in a Coq Call.

To recall the context:

- There was initially a discussion of documenting a restriction to expanding the CI.
- This restriction was justified by a lack of resources (both manpower to prepare overlays and machines to run the CI).
- The latter restriction is getting worked on by adding some external runners provided by Andres (possibly other runners could be provided by other important users).
- The restriction on the manpower was going to be lifted by asking project maintainers to do a bit more work.

But we also did not intend this to be such a radical change to the policy. Just a way to get rid of actually unmaintained projects in our CI, and to work with active maintainers of other projects to make sure that we keep supporting them without preventing Coq to evolve.

I like the idea of monorepos and expanding Coq's CI to better match the "Coq universe", but I believe that the comparison with monorepos in private companies is inadequate. Andres can certainly comment on that, but in the case of private companies, the head of the company decides which project is maintained and which project is not. Thus, if you want to change a library on which many projects depend, it makes sense that you have to adapt all the projects that depend on it. You will not remove projects just because they are not being ported. Here, we are talking about a universe of Coq projects which is maintained in a distributed way, where anyone can add a new item, but this should not mean that anytime this happens this increases the load on the shoulders of Coq developers, and that this should prohibit Coq from making breaking changes.

**Comment by ejgallego** [comment]
> Andres can certainly comment on that, but in the case of private companies, the head of the company decides which project is maintained and which project is not. Thus, if you want to change a library on which many projects depend, it makes sense that you have to adapt all the projects that depend on it. You will not remove projects just because they are not being ported. Here, we are talking about a universe of Coq projects which is maintained in a distributed way, where anyone can add a new item, but this should not mean that anytime this happens this increases the load on the shoulders of Coq developers, and that this should prohibit Coq from making breaking changes.

Very good points @Zimmi48 , thanks. What I know of private companies (not much tho) is a bit more complex, as usually you have different managerial levels, and not always monorepos, etc... Moreover, things in industry often go very very wrong haha.

My line of thought in the current situation is a bit more abstract and indeed heavily influenced by what I know about some large developments out there (math-comp and for example Jane Street libraries). In particular, I like to assume the following "mantras":

- effective software and proof development requires co-development and continuous, breaking refactoring
- developer experience and productivity is heavily influenced by tooling, review setup, build times, etc...
- developer behavior is not always rational, and often soft policies will not work and lead to creep (for example allowing warnings in builds)
- when a breaking change occurs (for example A, which depends on B, is broken by a change in B), cost of fixing the divergence grows _exponentially_ with time.

I do believe all these factors to hold in the extended setup of Coq + Coq's core users (let's say CI for now).

So in view of these factors, how do you reconcile point 1 with point 4? There are many possibilities, but indeed, a monorepos is attractive because it usually removes the kind of scary risk that come with "exponential cost". Moreover, if you allow software to diverge, poses a "pacing" problem: you need to be sure that the resources users have match the resources devs have, or they are outpaced, and thus diverging forever.

Also, monorepos tend to greatly help with point 2, but see my last point below.

I agree that the current model penalizes breaking compat as it imposes costs on the Coq devs. But note that this _by design_. In a sense, we choose "better safe and slow than sorry". Or said in other ways, we only break things if we can provide some reasonable witness of a fix.

There is another strong assumption in the original intent of the CI that is that in order to test Coq, you need to have a large codebase working with it. This is still pretty much true as of today, and IMHO an important need. Imagine contribs in CI start lagging, for example IRIS stops to compile. Then how do you test improvements pertinent to them? The OCaml compiler has suffered from this problem for example, but the context of Coq is IMHO different and the cost structure way more punishing.

Another (and last) important assumption that should be in the above list is the advantages of the cost model of distributed / multi-repos development vs mono-repos. If they are larger than for example the cost of divergence, then multi-repos setup is good.

**Comment by Zimmi48** [comment]
Thanks for clarifying your views. I can see that we are largely in agreement here, so the question with this PR is more to word it in a way that correctly capture the philosophy that we are defending:

- We need Coq to make breaking changes to evolve.
- We need a large Coq codebase (and as diverse as possible) to evolve jointly with Coq to be sure that we have enough to test Coq.
- We are happy to help the community evolve at the same pace as Coq.
- But we need to optimize the use of our resources while doing so.
- And thus we need to accept that some abandoned Coq code will die in the process.

Clearly, the goal is not to make acceptable that *actively maintained* Coq code should have to lag behind.

**Comment by ejgallego** [comment]
Thanks again for the points, a couple of comments:

> - We need Coq to make breaking changes to evolve.
> - We need a large Coq codebase (and as diverse as possible) to evolve jointly with Coq to be sure that we have enough to test Coq.

Indeed I think we could explore refining point 1 a to better reflect co-development:

- The goal of Coq is to make formal proofs a straightforward experience, this often means _joint_ refactorings between Coq code and proofs, for example, to adapt for better behaving tactics, or libraries.

Note that it is very clear to me personally (an opinion) that co-development is the only way to do formal proofs right, the tech we have is still very primitive, and IMHO we have a long way until we arrive to the provers of the future which will be NL-based. But not everyone shares this opinion, there are people that prefer a more classical development model, but I believe that this is too limiting.

As to give some perpestive on how things are in other projects, for example mathlib is on `leanprover/lean4:v4.12.0-rc1` as of today. So they are even not using stable releases due to the points above.

In a sense I feel that we have having trouble evaluating breaking changes in Coq. Ideally, the management layer of Coq would examine and survey the amount of breaking changes we had, and try to understand whether this changes have been justified.

It is my impression that in Coq we have often broken things without providing clear advantages to downstream users, and except for math-comp, we have not been doing co-development a lot. But indeed, this would deserve more analysis to see what the reality are.

I think a crutial point about breaking changes in other projects, is that these are often decided by the managerial layer, but here we have been way more informal.

> And thus we need to accept that some abandoned Coq code will die in the process.

Indeed no one is talking about preservation of completed projects, which is a different topic IMHO to be handled by people working already on code archive.

What I have in mind is our big, active, most representative projects.

**Comment by ejgallego** [comment]
> > I think a crutial point about breaking changes in other projects, is that these are often decided by the managerial layer, but here we have been way more informal.
> 
> If this helps, I would personally be OK with writing down in this policy that _all_ breaking changes need to be approved in a Coq Call (after the extent of breakage has been assessed). It probably wouldn't delay most changes, but @SkySkimmer or @ppedrot can probably comment on that (although likely not right now).

That could be interesting, however in my experience you want to manage these kind of impactful changes with a bit more structure, for example in relation with the Roadmap, etc... I am not sure the Coq call can replace some form of development / technical management. In the end, most choices systems like Coq face are (IMHO) _design_, not technical choices, so coherence tends to triumph.

Note also that I am assuming large changes, on the scale for example of Lean3 to Lean4 migration.

**Comment by Zimmi48** [comment]
We still need a process to be able to do smaller scale breaking changes.

I have been focused on the renaming lately, but I would like to go back to the roadmap effort after that, and follow up on Maxime's idea to use the Coq Calls as a way to check progress on the [short-term roadmap](https://github.com/coq/ceps/pull/69) (which itself should be aligned with the [long-term roadmap](https://rocq-prover.org/roadmap.html)). The current location of the short-term roadmap has not been satisfactory (not adapted to constant change), and this is why I'm experimenting with using a [GitHub project](https://github.com/orgs/coq/projects/28) with "iterations" for the renaming. If this works out, this is what I would like to use more generally for the short-term roadmap in the future.

**Comment by proux01** [comment]
> My line of thought in the current situation is a bit more abstract and indeed heavily influenced by what I know about some large developments out there (math-comp and for example Jane Street libraries).

Since you mention math-comp and I'm involved both there and here, let me state my understanding of the differences in CI handling:
* like here, math-comp has an extensive CI with reverse dependency testing (based on Nix), testing of course whatever reverse dependencies of math-comp is in Coq CI (that is mostly odd-order, fourcolor and analysis) and more
* unlike here, the CI policy in math-comp is almost non existent (or at least not written) so in some sense we lag behind Coq with that respect
* unlike here, things go well without much policy there because (my personal opinion and understanding of the situation):
  * like here, breakages of reverse dependencies are not so common
  * unlike here, when they happen, it's often the case that the author of the breaking PR and the broken reverse dependency are either the same person or very close
  * like here, de facto overlays are often written by the PR author but the above point and their small number usually makes this an acceptable overload
  * unlike here, the team is small enough that we can know / remember quirks like "dev X currently suffers a universe failure on master, person Y is working on it, it's fine to ignore and merge for now" (fortunately occurences are pretty rare)
  * unlike here, it's easy to commonly agree to remove something from CI in case it falls unmaintained (very rare though)

All in all it seems things are working in math-comp because it is a relatively small integrated team of people that are close enough, in some sense closer from the monorepo / monoteam world, so I'm not sure the comparison with Coq is really useful. Translated to the Coq team, that would almost mean Coq core team developing themselves everything in the CI, which is obviously not the case (neither something we should be aiming at IMHO). So again the repo vs team point.

**Comment by proux01** [comment]
> @proux01 , sorry I was not clear, I was not referring to the state of math-comp today, but I was thinking of their experience in the period of Feith-Thompsom

Thanks for the precision, I wasn't there at the time. I guess it was monorepo then. And I can imagine how not having Coq CI could have been a real pain (every now and then we still fall on some old comment mentioning somethings that broke in a particular version of Coq).

> when math-comp was actively developed.

FWIW math-comp is still actively developed ;-)

**Comment by ejgallego** [comment]
> FWIW math-comp is still actively developed ;-)

I meant Feit-Thompson, sorry!

But indeed, I understand this was a period where a lot of big things were developed, and very large refactoring happened. Most of what I know is via the papers and conversation with colleagues, mostly GG. It is a pity we don't have a better "post-morten" of the FT development effort.

> Thanks for the precision, I wasn't there at the time. I guess it was monorepo then. And I can imagine how not having Coq CI could have been a real pain (every now and then we still fall on some old comment mentioning somethings that broke in a particular version of Coq).

Yes, it was a monorepos _both_ for the plugin and the tactics. So for example, updates to tactics where synchronous.

My interest in their development process is because it is IMHO the large-scale proof effort that has produced a most polished result so far. So I wonder what kind of tools, methods, and setups, are required for that.

**Comment by gares** [comment]
> Yes, it was a monorepos _both_ for the plugin and the tactics. So for example, updates to tactics where synchronous.

Even back then, changes to the SSR tactics had almost zero backward incompatibilities.
New tactic features were typically applied in bulk refactorings, as the introduction of new notations and the like.
I think this happened more due to the way people were working than technical constraints, e.g. some people prefer large (visible) changes and to be sure their change makes sense globally before publishing them.

During the FT development timeframe, coq master (as well as the two previous branches) was tracked daily for both performance and functionality regressions by the MC team (via a self hosted jenkins). Most problems were due to changes in evarconv and low level tactic layers. Both functional and perf regressions happened, but not too often, also because the development of Coq was way less frenetic than today.

**Comment by proux01** [comment]
I tried to update this according to yesterday's Coq Call discussion. Please update it if I forgot something.

**Comment by Zimmi48** [comment]
Thanks for the update! Unfortunately, no one has posted notes from yesterday's Coq Call, so I'm using my memory, but I may be missing some aspects that we discussed. Two aspects that we highlighted needed clarification and that could be emphasized more are:

- We will disable the projects that have not been ported yet to not prevent merging the PR. The projects can then be re-enabled as soon as they are ported, but they risk hitting other breaking changes if their maintainers wait for too long.
- The 7-day delay is the minimal timeframe between the assignee notifying the project maintainers and the merge of the PR. Many reasons can justify longer delays (including having a maintainer ask for an extension).
- For some projects that we consider flagship projects (but which we don't have to list in our policy), we will not disable the project even if the delay has passed, but rather work with the maintainer to get their project updated.

**Comment by proux01** [comment]
>     * We will disable the projects that have not been ported yet to not prevent merging the PR. The projects can then be re-enabled as soon as they are ported, but they risk hitting other breaking changes if their maintainers wait for too long.

I added a precision.
 
>     * The 7-day delay is the minimal timeframe between the assignee notifying the project maintainers and the merge of the PR. Many reasons can justify longer delays (including having a maintainer ask for an extension).

Seems obvious to me since you can request extensions, but feel free to add a remark if you find an efficient way to state it.
 
>     * For some projects that we consider flagship projects (but which we don't have to list in our policy), we will not disable the project even if the delay has passed, but rather work with the maintainer to get their project updated.

Sure, we would never disable the Stdlib for instance, since it seats at the very bottom of the hierarchy it is likely any change to it would get most, if not all, of its patch prepared by the PR author as part of the first step

>    * you must assess the impact of the breakage and porting
     effort required on a few CI entries that exercise the
     functionality being changed.

**Comment by Zimmi48** [comment]
For me too, this was all obvious, but what the discussion highlighted is that some people would feel more comfortable if this is written down. I can push to this branch, but then someone else will need to be the assignee.

**Comment by proux01** [comment]
Ok, I added a paragraph to state the obvious. Anyone feel free to improve the wording.

**Comment by Zimmi48** [comment]
If there are no more comments, I intend to merge by the end of the week.

**Pull request merged**
